### PR TITLE
feat: Add deterministic guardrails filtering action

### PR DESCRIPTION
**Pull request by apetraru-uipath**
# Filter Action Implementation - PR Summary

## Overview
Implements filter action for deterministic tool guardrails, allowing selective removal of input/output fields based on guardrail rules.

### ✅ Execution Stage Awareness
- **PRE_EXECUTION**: Filters input fields only
- **POST_EXECUTION**: Filters output fields only
- Guardrails trigger based on rules, actions apply based on stage

### ✅ Python Dict String Handling
- Added `ast.literal_eval()` fallback for tool outputs stored as dict string representations
- Handles JSON strings, native dicts, and Python literal strings

### ✅ Type Safety & Code Quality
- Fixed mypy type errors with proper `BaseGuardrail` annotations
- Resolved test file naming conflicts (`test_utils.py` → `test_guardrails_utils.py`)
- Replaced mock objects with proper `GuardrailSelector` types

### Modified Files
- `filter_action.py` - Core filtering logic with stage-aware field filtering
- `utils.py` - Enhanced data extraction with dict literal parsing
- `guardrails_factory.py` - Type safety improvements
- `guardrails_subgraph.py` - Bug fix for POST_EXECUTION node routing

### Test Coverage
- **17 comprehensive tests** (consolidated from 20, -15% reduction)
- **165 total guardrails tests** passing
- All edge cases covered: empty fields, missing data, non-JSON content, multiple fields

## Validation

Tested against 7 agent configurations covering:
- Always rules (PRE/POST)
- Conditional rules (word matching)
- Field selectors (specific/all)
- Same-field filtering
- Multiple field filtering

### CODE REVIEW

File: tests/agent/guardrails/test_guardrails_factory.py [hunk 1]
```diff
@@ -158,6 +161,198 @@ def test_escalate_action_is_mapped_with_app_and_recipient(self) -> None:
...
+
+        result = build_guardrails_with_actions([guardrail])
+
+        assert len(result) == 1
+        gr, action = result[0]
+        assert gr is guardrail
+        assert isinstance(action, FilterAction)
+        assert action.fields == []
+
+    def test_filter_action_with_empty_fields_list(self) -> None:
```
Comment (valentinabojan) [inline tests/agent/guardrails/test_guardrails_factory.py]: this test is same as test_filter_action_with_empty_fields. Only keep one of them

File: tests/agent/guardrails/actions/test_filter_action.py [hunk 2]
```diff
@@ -85,18 +98,631 @@ async def test_node_name_and_exception_for_unsupported_scopes(
...
+
+        # Input should be filtered at POST_EXECUTION
+        assert isinstance(result, Command)
+        assert result.update is not None
+        updated_ai_message = result.update["messages"][0]
+        assert updated_ai_message.tool_calls[0]["args"] == {}
+
+    # Test Agent6: Same-field filtering (input)
+    @pytest.mark.asyncio
+    async def test_same_field_filter_input(self) -> None:
```
Comment (valentinabojan) [inline tests/agent/guardrails/actions/test_filter_action.py]: This test seems identical to test_filter_both_input_and_output_at_pre

File: src/uipath_langchain/agent/guardrails/guardrails_factory.py [hunk 3]
```diff
@@ -233,6 +239,12 @@ def build_guardrails_with_actions(
             converted_guardrail = _convert_agent_custom_guardrail_to_deterministic(
                 guardrail
             )
+        else:
```
[the remark this answers is not in the document]
Comment (apetraru-uipath) [reply src/uipath_langchain/agent/guardrails/guardrails_factory.py]: Fixed
